engine/src/ui/canvas/draw_list.cpp

```diff
@@ -0,0 +1,122 @@
+#include <cstring>
+
+#include <cubos/core/log.hpp>
+
+#include <cubos/engine/ui/canvas/draw_list.hpp>
+
+cubos::engine::UIDrawList::UIDrawListEntryBuilder::UIDrawListEntryBuilder(UIDrawListEntry& entry)
+    : mEntry(entry)
+{
+}
+
+cubos::engine::UIDrawList::UIDrawListEntryBuilder cubos::engine::UIDrawList::UIDrawListEntryBuilder::withTexture(
+    int texIndex, cubos::core::gl::Texture2D texture, cubos::core::gl::Sampler sampler)
+{
+    if (texIndex >= UIDrawType::MaxTextures)
+    {
+        CUBOS_WARN("Invalid texture index, texture will be dropped");
+        return *this;
+    }
+    mEntry.command.samplers[texIndex] = sampler;
+    mEntry.command.textures[texIndex] = texture;
+    return *this;
+}
+
+cubos::engine::UIDrawList::UIDrawListEntry cubos::engine::UIDrawList::UIDrawListEntryBuilder::entry() const
+{
+    return mEntry;
+}
+
+cubos::engine::UIDrawList::UIDrawList()
+{
+    mData = operator new(sizeof(char) * mDataCapacity);
+}
+
+cubos::engine::UIDrawList::UIDrawList(UIDrawList&& other) noexcept
+    : mEntries(std::move(other.mEntries))
+    , mData(other.mData)
+    , mDataCapacity(other.mDataCapacity)
+    , mDataSize(other.mDataSize)
+{
+    other.mData = nullptr;
+    other.mDataSize = 0;
+    other.mDataCapacity = 0;
+}
+
+cubos::engine::UIDrawList::UIDrawList(const UIDrawList& other) noexcept
+    : mEntries(other.mEntries)
+    , mDataCapacity(other.mDataCapacity)
+    , mDataSize(other.mDataSize)
+{
+    mData = operator new(sizeof(char) * other.mDataSize);
+    std::memcpy(mData, other.mData, sizeof(char) * other.mDataSize);
+}
+
+cubos::engine::UIDrawList::~UIDrawList()
+{
+    operator delete(mData);
+}
+
+void cubos::engine::UIDrawList::push(const UIDrawType& type, const UIDrawCommand& command, const void* data)
+{
+    auto entry = UIDrawListEntry{type, command, mDataSize};
+    entry.offset = mDataSize;
+    mEntries.push_back(entry);
+    if (mDataSize + entry.type.perElementSize > mDataCapacity)
+    {
+        mDataCapacity += entry.type.perElementSize;
+        mDataCapacity *= 2;
+        void* temp = operator new(sizeof(char) * mDataCapacity);
+        std::memcpy(temp, mData, mDataSize);
+        operator delete(mData);
+        mData = temp;
+    }
+    std::memcpy(static_cast<char*>(mData) + mDataSize, data, entry.type.perElementSize);
+    mDataSize += entry.type.perElementSize;
```

Comment:
Can't you make this return a builder too? So that in the function below you can just call this one, to avoid duplicating this allocation logic.